NPE if the server doesn't provide any HTTP content for an error

A user reported an NPE on the [Gitter channel](gitter.im/google/jib).

```
Caused by: java.lang.NullPointerException
	at com.fasterxml.jackson.core.JsonFactory.createParser(JsonFactory.java:889)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3005)
	at com.google.cloud.tools.jib.json.JsonTemplateMapper.readJson(JsonTemplateMapper.java:118)
	at com.google.cloud.tools.jib.json.JsonTemplateMapper.readJson (JsonTemplateMapper.java:118)
	at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.newRegistryErrorException (RegistryEndpointCaller.java:194)
```

The NPE is when there was an error communicating with the server. Jib tries to parse the content of the error message (supposed to be a JSON) from the server.
```java
      ErrorResponseTemplate errorResponse =
          JsonTemplateMapper.readJson(responseException.getContent(), ErrorResponseTemplate.class);
```

I noticed that if we pass a null string, `JsonTemplateMapper.readJson()` throws NPE with the same stacktrace.
```java
JsonTemplateMapper.readJson((String) null, ErrorResponseTemplate.class);
```

Turns out `responseException.getContent()` can return null if there was no content from the server. The reason I think NullAway couldn't catch this is that the return value of `getContet()` basically comes from a method in the Google HTTP Client library.

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -1,24 +1,21 @@
 /*
  * Copyright 2018 Google LLC.
  *
- * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
- * in compliance with the License. You may obtain a copy of the License at
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
  *
- * http://www.apache.org/licenses/LICENSE-2.0
+ *      http://www.apache.org/licenses/LICENSE-2.0
  *
- * Unless required by applicable law or agreed to in writing, software distributed under the License
- * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
- * or implied. See the License for the specific language governing permissions and limitations under
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
  * the License.
  */
 
 package com.google.cloud.tools.jib.registry;
 
-import java.io.IOException;
-import java.net.URL;
-import java.util.Locale;
-import javax.annotation.Nullable;
-import javax.net.ssl.SSLException;
 import com.google.api.client.http.HttpStatusCodes;
 import com.google.cloud.tools.jib.api.InsecureRegistryException;
 import com.google.cloud.tools.jib.api.LogEvent;
@@ -35,6 +32,11 @@ import com.google.cloud.tools.jib.json.JsonTemplateMapper;
 import com.google.cloud.tools.jib.registry.json.ErrorEntryTemplate;
 import com.google.cloud.tools.jib.registry.json.ErrorResponseTemplate;
 import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Locale;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLException;
 
 /**
  * Makes requests to a registry endpoint.
@@ -164,6 +166,7 @@ class RegistryEndpointCaller<T> {
           throw responseException;
         }
       }
+      
 
     } catch (IOException ex) {
       logError("I/O error for image [" + serverUrl + "/" + imageName + "]:");

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -49,8 +49,7 @@ class RegistryEndpointCaller<T> {
    * <a href=
    * "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308">https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308</a>.
    */
-  @VisibleForTesting
-  static final int STATUS_CODE_PERMANENT_REDIRECT = 308;
+  @VisibleForTesting static final int STATUS_CODE_PERMANENT_REDIRECT = 308;
 
   // https://github.com/GoogleContainerTools/jib/issues/1316
   @VisibleForTesting
@@ -71,11 +70,9 @@ class RegistryEndpointCaller<T> {
   }
 
   private final EventHandlers eventHandlers;
-  @Nullable
-  private final String userAgent;
+  @Nullable private final String userAgent;
   private final RegistryEndpointProvider<T> registryEndpointProvider;
-  @Nullable
-  private final Authorization authorization;
+  @Nullable private final Authorization authorization;
   private final RegistryEndpointRequestProperties registryEndpointRequestProperties;
   private final FailoverHttpClient httpClient;
 
@@ -90,8 +87,11 @@ class RegistryEndpointCaller<T> {
    * @param httpClient HTTP client
    */
   @VisibleForTesting
-  RegistryEndpointCaller(EventHandlers eventHandlers, @Nullable String userAgent,
-      RegistryEndpointProvider<T> registryEndpointProvider, @Nullable Authorization authorization,
+  RegistryEndpointCaller(
+      EventHandlers eventHandlers,
+      @Nullable String userAgent,
+      RegistryEndpointProvider<T> registryEndpointProvider,
+      @Nullable Authorization authorization,
       RegistryEndpointRequestProperties registryEndpointRequestProperties,
       FailoverHttpClient httpClient) {
     this.eventHandlers = eventHandlers;
@@ -127,10 +127,13 @@ class RegistryEndpointCaller<T> {
     String serverUrl = registryEndpointRequestProperties.getServerUrl();
     String imageName = registryEndpointRequestProperties.getImageName();
 
-    Request.Builder requestBuilder = Request.builder().setUserAgent(userAgent)
-        .setHttpTimeout(JibSystemProperties.getHttpTimeout())
-        .setAccept(registryEndpointProvider.getAccept())
-        .setBody(registryEndpointProvider.getContent()).setAuthorization(authorization);
+    Request.Builder requestBuilder =
+        Request.builder()
+            .setUserAgent(userAgent)
+            .setHttpTimeout(JibSystemProperties.getHttpTimeout())
+            .setAccept(registryEndpointProvider.getAccept())
+            .setBody(registryEndpointProvider.getContent())
+            .setAuthorization(authorization);
 
     try (Response response =
         httpClient.call(registryEndpointProvider.getHttpMethod(), url, requestBuilder.build())) {
@@ -145,8 +148,8 @@ class RegistryEndpointCaller<T> {
       } catch (ResponseException responseException) {
         if (responseException.getStatusCode() == HttpStatusCodes.STATUS_CODE_BAD_REQUEST
             || responseException.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND
-            || responseException
-                .getStatusCode() == HttpStatusCodes.STATUS_CODE_METHOD_NOT_ALLOWED) {
+            || responseException.getStatusCode()
+                == HttpStatusCodes.STATUS_CODE_METHOD_NOT_ALLOWED) {
           // The name or reference was invalid.
           throw newRegistryErrorException(responseException);
 
@@ -166,7 +169,6 @@ class RegistryEndpointCaller<T> {
           throw responseException;
         }
       }
-      
 
     } catch (IOException ex) {
       logError("I/O error for image [" + serverUrl + "/" + imageName + "]:");
@@ -183,25 +185,30 @@ class RegistryEndpointCaller<T> {
 
   @VisibleForTesting
   RegistryErrorException newRegistryErrorException(ResponseException responseException) {
-    RegistryErrorExceptionBuilder registryErrorExceptionBuilder = new RegistryErrorExceptionBuilder(
-        registryEndpointProvider.getActionDescription(), responseException);
+    RegistryErrorExceptionBuilder registryErrorExceptionBuilder =
+        new RegistryErrorExceptionBuilder(
+            registryEndpointProvider.getActionDescription(), responseException);
     if (responseException.getContent() != null) {
       try {
-        ErrorResponseTemplate errorResponse = JsonTemplateMapper
-            .readJson(responseException.getContent(), ErrorResponseTemplate.class);
+        ErrorResponseTemplate errorResponse =
+            JsonTemplateMapper.readJson(
+                responseException.getContent(), ErrorResponseTemplate.class);
         for (ErrorEntryTemplate errorEntry : errorResponse.getErrors()) {
           registryErrorExceptionBuilder.addReason(errorEntry);
         }
       } catch (IOException ex) {
-        registryErrorExceptionBuilder.addReason("registry returned error code "
-            + responseException.getStatusCode()
-            + "; possible causes include invalid or wrong reference. Actual error output follows:\n"
-            + responseException.getContent() + "\n");
+        registryErrorExceptionBuilder.addReason(
+            "registry returned error code "
+                + responseException.getStatusCode()
+                + "; possible causes include invalid or wrong reference. Actual error output follows:\n"
+                + responseException.getContent()
+                + "\n");
       }
     } else {
-      registryErrorExceptionBuilder.addReason("registry returned error code "
-          + responseException.getStatusCode()
-          + "; possible causes include invalid or wrong reference \n");
+      registryErrorExceptionBuilder.addReason(
+          "registry returned error code "
+              + responseException.getStatusCode()
+              + "; possible causes include invalid or wrong reference \n");
     }
     return registryErrorExceptionBuilder.build();
   }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -1,21 +1,24 @@
 /*
  * Copyright 2018 Google LLC.
  *
- * Licensed under the Apache License, Version 2.0 (the "License"); you may not
- * use this file except in compliance with the License. You may obtain a copy of
- * the License at
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
  *
- *      http://www.apache.org/licenses/LICENSE-2.0
+ * http://www.apache.org/licenses/LICENSE-2.0
  *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
- * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
- * License for the specific language governing permissions and limitations under
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
  * the License.
  */
 
 package com.google.cloud.tools.jib.registry;
 
+import java.io.IOException;
+import java.net.URL;
+import java.util.Locale;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLException;
 import com.google.api.client.http.HttpStatusCodes;
 import com.google.cloud.tools.jib.api.InsecureRegistryException;
 import com.google.cloud.tools.jib.api.LogEvent;
@@ -32,11 +35,6 @@ import com.google.cloud.tools.jib.json.JsonTemplateMapper;
 import com.google.cloud.tools.jib.registry.json.ErrorEntryTemplate;
 import com.google.cloud.tools.jib.registry.json.ErrorResponseTemplate;
 import com.google.common.annotations.VisibleForTesting;
-import java.io.IOException;
-import java.net.URL;
-import java.util.Locale;
-import javax.annotation.Nullable;
-import javax.net.ssl.SSLException;
 
 /**
  * Makes requests to a registry endpoint.
@@ -46,10 +44,11 @@ import javax.net.ssl.SSLException;
 class RegistryEndpointCaller<T> {
 
   /**
-   * <a
-   * href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308">https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308</a>.
+   * <a href=
+   * "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308">https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308</a>.
    */
-  @VisibleForTesting static final int STATUS_CODE_PERMANENT_REDIRECT = 308;
+  @VisibleForTesting
+  static final int STATUS_CODE_PERMANENT_REDIRECT = 308;
 
   // https://github.com/GoogleContainerTools/jib/issues/1316
   @VisibleForTesting
@@ -70,9 +69,11 @@ class RegistryEndpointCaller<T> {
   }
 
   private final EventHandlers eventHandlers;
-  @Nullable private final String userAgent;
+  @Nullable
+  private final String userAgent;
   private final RegistryEndpointProvider<T> registryEndpointProvider;
-  @Nullable private final Authorization authorization;
+  @Nullable
+  private final Authorization authorization;
   private final RegistryEndpointRequestProperties registryEndpointRequestProperties;
   private final FailoverHttpClient httpClient;
 
@@ -87,11 +88,8 @@ class RegistryEndpointCaller<T> {
    * @param httpClient HTTP client
    */
   @VisibleForTesting
-  RegistryEndpointCaller(
-      EventHandlers eventHandlers,
-      @Nullable String userAgent,
-      RegistryEndpointProvider<T> registryEndpointProvider,
-      @Nullable Authorization authorization,
+  RegistryEndpointCaller(EventHandlers eventHandlers, @Nullable String userAgent,
+      RegistryEndpointProvider<T> registryEndpointProvider, @Nullable Authorization authorization,
       RegistryEndpointRequestProperties registryEndpointRequestProperties,
       FailoverHttpClient httpClient) {
     this.eventHandlers = eventHandlers;
@@ -127,13 +125,10 @@ class RegistryEndpointCaller<T> {
     String serverUrl = registryEndpointRequestProperties.getServerUrl();
     String imageName = registryEndpointRequestProperties.getImageName();
 
-    Request.Builder requestBuilder =
-        Request.builder()
-            .setUserAgent(userAgent)
-            .setHttpTimeout(JibSystemProperties.getHttpTimeout())
-            .setAccept(registryEndpointProvider.getAccept())
-            .setBody(registryEndpointProvider.getContent())
-            .setAuthorization(authorization);
+    Request.Builder requestBuilder = Request.builder().setUserAgent(userAgent)
+        .setHttpTimeout(JibSystemProperties.getHttpTimeout())
+        .setAccept(registryEndpointProvider.getAccept())
+        .setBody(registryEndpointProvider.getContent()).setAuthorization(authorization);
 
     try (Response response =
         httpClient.call(registryEndpointProvider.getHttpMethod(), url, requestBuilder.build())) {
@@ -148,8 +143,8 @@ class RegistryEndpointCaller<T> {
       } catch (ResponseException responseException) {
         if (responseException.getStatusCode() == HttpStatusCodes.STATUS_CODE_BAD_REQUEST
             || responseException.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND
-            || responseException.getStatusCode()
-                == HttpStatusCodes.STATUS_CODE_METHOD_NOT_ALLOWED) {
+            || responseException
+                .getStatusCode() == HttpStatusCodes.STATUS_CODE_METHOD_NOT_ALLOWED) {
           // The name or reference was invalid.
           throw newRegistryErrorException(responseException);
 
@@ -185,25 +180,26 @@ class RegistryEndpointCaller<T> {
 
   @VisibleForTesting
   RegistryErrorException newRegistryErrorException(ResponseException responseException) {
-    RegistryErrorExceptionBuilder registryErrorExceptionBuilder =
-        new RegistryErrorExceptionBuilder(
-            registryEndpointProvider.getActionDescription(), responseException);
-
-    try {
-      ErrorResponseTemplate errorResponse =
-          JsonTemplateMapper.readJson(responseException.getContent(), ErrorResponseTemplate.class);
-      for (ErrorEntryTemplate errorEntry : errorResponse.getErrors()) {
-        registryErrorExceptionBuilder.addReason(errorEntry);
+    RegistryErrorExceptionBuilder registryErrorExceptionBuilder = new RegistryErrorExceptionBuilder(
+        registryEndpointProvider.getActionDescription(), responseException);
+    if (responseException.getContent() != null) {
+      try {
+        ErrorResponseTemplate errorResponse = JsonTemplateMapper
+            .readJson(responseException.getContent(), ErrorResponseTemplate.class);
+        for (ErrorEntryTemplate errorEntry : errorResponse.getErrors()) {
+          registryErrorExceptionBuilder.addReason(errorEntry);
+        }
+      } catch (IOException ex) {
+        registryErrorExceptionBuilder.addReason("registry returned error code "
+            + responseException.getStatusCode()
+            + "; possible causes include invalid or wrong reference. Actual error output follows:\n"
+            + responseException.getContent() + "\n");
       }
-    } catch (IOException ex) {
-      registryErrorExceptionBuilder.addReason(
-          "registry returned error code "
-              + responseException.getStatusCode()
-              + "; possible causes include invalid or wrong reference. Actual error output follows:\n"
-              + responseException.getContent()
-              + "\n");
+    } else {
+      registryErrorExceptionBuilder.addReason("registry returned error code "
+          + responseException.getStatusCode()
+          + "; possible causes include invalid or wrong reference \n");
     }
-
     return registryErrorExceptionBuilder.build();
   }
 

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -208,7 +208,7 @@ class RegistryEndpointCaller<T> {
       registryErrorExceptionBuilder.addReason(
           "registry returned error code "
               + responseException.getStatusCode()
-              + "; but did not return any details; possible causes include invalid or wrong reference, or proxy/firewall/VPN interfering \n");
+              + " but did not return any details; possible causes include invalid or wrong reference, or proxy/firewall/VPN interfering \n");
     }
     return registryErrorExceptionBuilder.build();
   }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -46,8 +46,7 @@ import javax.net.ssl.SSLException;
 class RegistryEndpointCaller<T> {
 
   /**
-   * <a href=
-   * "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308">https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308</a>.
+   * <a href = "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308">https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308</a>.
    */
   @VisibleForTesting static final int STATUS_CODE_PERMANENT_REDIRECT = 308;
 
@@ -208,7 +207,7 @@ class RegistryEndpointCaller<T> {
       registryErrorExceptionBuilder.addReason(
           "registry returned error code "
               + responseException.getStatusCode()
-              + "; possible causes include invalid or wrong reference \n");
+              + "; but did not return any details; possible causes include invalid or wrong reference, or proxy/firewall/VPN interfering \n");
     }
     return registryErrorExceptionBuilder.build();
   }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -46,7 +46,8 @@ import javax.net.ssl.SSLException;
 class RegistryEndpointCaller<T> {
 
   /**
-   * <a href = "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308">https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308</a>.
+   * <a href =
+   * "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308">https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308</a>.
    */
   @VisibleForTesting static final int STATUS_CODE_PERMANENT_REDIRECT = 308;
 

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -446,7 +446,7 @@ public class RegistryEndpointCallerTest {
         endpointCaller.newRegistryErrorException(httpException);
     Assert.assertSame(httpException, registryException.getCause());
     Assert.assertEquals(
-        "Tried to actionDescription but failed because: registry returned error code 404; "
+        "Tried to actionDescription but failed because: registry returned error code 404 "
             + "but did not return any details; possible causes include invalid or wrong reference, or proxy/firewall/VPN interfering \n"
             + " | If this is a bug, please file an issue at "
             + "https://github.com/GoogleContainerTools/jib/issues/new",

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -436,7 +436,7 @@ public class RegistryEndpointCallerTest {
   }
 
   @Test
-  public void testNewRegistryErrorException_nullJsonErrorOutput() {
+  public void testNewRegistryErrorException_noOutputFromRegistry() {
     ResponseException httpException = Mockito.mock(ResponseException.class);
     // Registry returning null error output
     Mockito.when(httpException.getContent()).thenReturn(null);
@@ -447,7 +447,7 @@ public class RegistryEndpointCallerTest {
     Assert.assertSame(httpException, registryException.getCause());
     Assert.assertEquals(
         "Tried to actionDescription but failed because: registry returned error code 404; "
-            + "possible causes include invalid or wrong reference \n"
+            + "but did not return any details; possible causes include invalid or wrong reference, or proxy/firewall/VPN interfering \n"
             + " | If this is a bug, please file an issue at "
             + "https://github.com/GoogleContainerTools/jib/issues/new",
         registryException.getMessage());

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -16,21 +16,6 @@
 
 package com.google.cloud.tools.jib.registry;
 
-import com.google.api.client.http.HttpStatusCodes;
-import com.google.cloud.tools.jib.api.InsecureRegistryException;
-import com.google.cloud.tools.jib.api.LogEvent;
-import com.google.cloud.tools.jib.api.RegistryException;
-import com.google.cloud.tools.jib.api.RegistryUnauthorizedException;
-import com.google.cloud.tools.jib.event.EventHandlers;
-import com.google.cloud.tools.jib.global.JibSystemProperties;
-import com.google.cloud.tools.jib.http.Authorization;
-import com.google.cloud.tools.jib.http.BlobHttpContent;
-import com.google.cloud.tools.jib.http.FailoverHttpClient;
-import com.google.cloud.tools.jib.http.Request;
-import com.google.cloud.tools.jib.http.RequestWrapper;
-import com.google.cloud.tools.jib.http.Response;
-import com.google.cloud.tools.jib.http.ResponseException;
-import com.google.common.io.CharStreams;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStreamReader;
@@ -55,6 +40,21 @@ import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.mockito.Mockito;
 import org.mockito.junit.MockitoJUnitRunner;
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.cloud.tools.jib.api.InsecureRegistryException;
+import com.google.cloud.tools.jib.api.LogEvent;
+import com.google.cloud.tools.jib.api.RegistryException;
+import com.google.cloud.tools.jib.api.RegistryUnauthorizedException;
+import com.google.cloud.tools.jib.event.EventHandlers;
+import com.google.cloud.tools.jib.global.JibSystemProperties;
+import com.google.cloud.tools.jib.http.Authorization;
+import com.google.cloud.tools.jib.http.BlobHttpContent;
+import com.google.cloud.tools.jib.http.FailoverHttpClient;
+import com.google.cloud.tools.jib.http.Request;
+import com.google.cloud.tools.jib.http.RequestWrapper;
+import com.google.cloud.tools.jib.http.Response;
+import com.google.cloud.tools.jib.http.ResponseException;
+import com.google.common.io.CharStreams;
 
 /** Tests for {@link RegistryEndpointCaller}. */
 @RunWith(MockitoJUnitRunner.class)
@@ -434,6 +434,25 @@ public class RegistryEndpointCallerTest {
             + "https://github.com/GoogleContainerTools/jib/issues/new",
         registryException.getMessage());
   }
+  
+
+  @Test
+  public void testNewRegistryErrorException_nullJsonErrorOutput() {
+    ResponseException httpException = Mockito.mock(ResponseException.class);
+    // Registry returning null error output
+    Mockito.when(httpException.getContent()).thenReturn(null);
+    Mockito.when(httpException.getStatusCode()).thenReturn(404);
+
+    RegistryErrorException registryException =
+        endpointCaller.newRegistryErrorException(httpException);
+    Assert.assertSame(httpException, registryException.getCause());
+    Assert.assertEquals(
+        "Tried to actionDescription but failed because: registry returned error code 404; "
+            + "possible causes include invalid or wrong reference \n"
+            +" | If this is a bug, please file an issue at "
+            + "https://github.com/GoogleContainerTools/jib/issues/new",
+        registryException.getMessage());
+  }
 
   /**
    * Verifies that a response with {@code httpStatusCode} throws {@link
@@ -477,6 +496,7 @@ public class RegistryEndpointCallerTest {
     }
   }
 
+  
   private void setUpRegistryResponse(Exception exceptionToThrow)
       throws MalformedURLException, IOException {
     Mockito.when(

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -16,6 +16,21 @@
 
 package com.google.cloud.tools.jib.registry;
 
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.cloud.tools.jib.api.InsecureRegistryException;
+import com.google.cloud.tools.jib.api.LogEvent;
+import com.google.cloud.tools.jib.api.RegistryException;
+import com.google.cloud.tools.jib.api.RegistryUnauthorizedException;
+import com.google.cloud.tools.jib.event.EventHandlers;
+import com.google.cloud.tools.jib.global.JibSystemProperties;
+import com.google.cloud.tools.jib.http.Authorization;
+import com.google.cloud.tools.jib.http.BlobHttpContent;
+import com.google.cloud.tools.jib.http.FailoverHttpClient;
+import com.google.cloud.tools.jib.http.Request;
+import com.google.cloud.tools.jib.http.RequestWrapper;
+import com.google.cloud.tools.jib.http.Response;
+import com.google.cloud.tools.jib.http.ResponseException;
+import com.google.common.io.CharStreams;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStreamReader;
@@ -40,21 +55,6 @@ import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.mockito.Mockito;
 import org.mockito.junit.MockitoJUnitRunner;
-import com.google.api.client.http.HttpStatusCodes;
-import com.google.cloud.tools.jib.api.InsecureRegistryException;
-import com.google.cloud.tools.jib.api.LogEvent;
-import com.google.cloud.tools.jib.api.RegistryException;
-import com.google.cloud.tools.jib.api.RegistryUnauthorizedException;
-import com.google.cloud.tools.jib.event.EventHandlers;
-import com.google.cloud.tools.jib.global.JibSystemProperties;
-import com.google.cloud.tools.jib.http.Authorization;
-import com.google.cloud.tools.jib.http.BlobHttpContent;
-import com.google.cloud.tools.jib.http.FailoverHttpClient;
-import com.google.cloud.tools.jib.http.Request;
-import com.google.cloud.tools.jib.http.RequestWrapper;
-import com.google.cloud.tools.jib.http.Response;
-import com.google.cloud.tools.jib.http.ResponseException;
-import com.google.common.io.CharStreams;
 
 /** Tests for {@link RegistryEndpointCaller}. */
 @RunWith(MockitoJUnitRunner.class)
@@ -87,8 +87,8 @@ public class RegistryEndpointCallerTest {
     @Nullable
     @Override
     public String handleResponse(Response response) throws IOException {
-      return CharStreams.toString(
-          new InputStreamReader(response.getBody(), StandardCharsets.UTF_8));
+      return CharStreams
+          .toString(new InputStreamReader(response.getBody(), StandardCharsets.UTF_8));
     }
 
     @Override
@@ -109,24 +109,23 @@ public class RegistryEndpointCallerTest {
     return mock;
   }
 
-  @Rule public final RestoreSystemProperties systemPropertyRestorer = new RestoreSystemProperties();
+  @Rule
+  public final RestoreSystemProperties systemPropertyRestorer = new RestoreSystemProperties();
 
-  @Mock private EventHandlers mockEventHandlers;
-  @Mock private FailoverHttpClient mockHttpClient;
-  @Mock private Response mockResponse;
+  @Mock
+  private EventHandlers mockEventHandlers;
+  @Mock
+  private FailoverHttpClient mockHttpClient;
+  @Mock
+  private Response mockResponse;
 
   private RegistryEndpointCaller<String> endpointCaller;
 
   @Before
   public void setUp() throws IOException {
-    endpointCaller =
-        new RegistryEndpointCaller<>(
-            mockEventHandlers,
-            "userAgent",
-            new TestRegistryEndpointProvider(),
-            Authorization.fromBasicCredentials("user", "pass"),
-            new RegistryEndpointRequestProperties("serverUrl", "imageName"),
-            mockHttpClient);
+    endpointCaller = new RegistryEndpointCaller<>(mockEventHandlers, "userAgent",
+        new TestRegistryEndpointProvider(), Authorization.fromBasicCredentials("user", "pass"),
+        new RegistryEndpointRequestProperties("serverUrl", "imageName"), mockHttpClient);
 
     Mockito.when(mockResponse.getBody())
         .thenReturn(new ByteArrayInputStream("body".getBytes(StandardCharsets.UTF_8)));
@@ -249,9 +248,8 @@ public class RegistryEndpointCallerTest {
 
     } catch (IOException ex) {
       Assert.assertSame(ioException, ex);
-      Mockito.verify(mockEventHandlers)
-          .dispatch(
-              LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
+      Mockito.verify(mockEventHandlers).dispatch(
+          LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    java.io.IOException\u001B[0m"));
       Mockito.verify(mockEventHandlers)
@@ -271,19 +269,17 @@ public class RegistryEndpointCallerTest {
 
     } catch (IOException ex) {
       Assert.assertSame(ioException, ex);
-      Mockito.verify(mockEventHandlers)
-          .dispatch(
-              LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
+      Mockito.verify(mockEventHandlers).dispatch(
+          LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    java.io.IOException\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    this is due to broken pipe\u001B[0m"));
       Mockito.verify(mockEventHandlers)
-          .dispatch(
-              LogEvent.error(
-                  "\u001B[31;1mbroken pipe: the server shut down the connection. Check the server "
-                      + "log if possible. This could also be a proxy issue. For example, a proxy "
-                      + "may prevent sending packets that are too large.\u001B[0m"));
+          .dispatch(LogEvent.error(
+              "\u001B[31;1mbroken pipe: the server shut down the connection. Check the server "
+                  + "log if possible. This could also be a proxy issue. For example, a proxy "
+                  + "may prevent sending packets that are too large.\u001B[0m"));
       Mockito.verifyNoMoreInteractions(mockEventHandlers);
     }
   }
@@ -297,9 +293,8 @@ public class RegistryEndpointCallerTest {
       Assert.fail();
 
     } catch (IOException ex) {
-      Mockito.verify(mockEventHandlers)
-          .dispatch(
-              LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
+      Mockito.verify(mockEventHandlers).dispatch(
+          LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    java.io.IOException\u001B[0m"));
       Mockito.verify(mockEventHandlers)
@@ -403,9 +398,8 @@ public class RegistryEndpointCallerTest {
   @Test
   public void testNewRegistryErrorException_jsonErrorOutput() {
     ResponseException httpException = Mockito.mock(ResponseException.class);
-    Mockito.when(httpException.getContent())
-        .thenReturn(
-            "{\"errors\": [{\"code\": \"MANIFEST_UNKNOWN\", \"message\": \"manifest unknown\"}]}");
+    Mockito.when(httpException.getContent()).thenReturn(
+        "{\"errors\": [{\"code\": \"MANIFEST_UNKNOWN\", \"message\": \"manifest unknown\"}]}");
 
     RegistryErrorException registryException =
         endpointCaller.newRegistryErrorException(httpException);
@@ -429,12 +423,11 @@ public class RegistryEndpointCallerTest {
     Assert.assertEquals(
         "Tried to actionDescription but failed because: registry returned error code 404; "
             + "possible causes include invalid or wrong reference. Actual error output follows:\n"
-            + ">>>>> (404) page not found <<<<<\n"
-            + " | If this is a bug, please file an issue at "
+            + ">>>>> (404) page not found <<<<<\n" + " | If this is a bug, please file an issue at "
             + "https://github.com/GoogleContainerTools/jib/issues/new",
         registryException.getMessage());
   }
-  
+
 
   @Test
   public void testNewRegistryErrorException_nullJsonErrorOutput() {
@@ -449,14 +442,14 @@ public class RegistryEndpointCallerTest {
     Assert.assertEquals(
         "Tried to actionDescription but failed because: registry returned error code 404; "
             + "possible causes include invalid or wrong reference \n"
-            +" | If this is a bug, please file an issue at "
+            + " | If this is a bug, please file an issue at "
             + "https://github.com/GoogleContainerTools/jib/issues/new",
         registryException.getMessage());
   }
 
   /**
-   * Verifies that a response with {@code httpStatusCode} throws {@link
-   * RegistryUnauthorizedException}.
+   * Verifies that a response with {@code httpStatusCode} throws
+   * {@link RegistryUnauthorizedException}.
    */
   private void verifyThrowsRegistryUnauthorizedException(int httpStatusCode)
       throws IOException, RegistryException {
@@ -474,8 +467,8 @@ public class RegistryEndpointCallerTest {
   }
 
   /**
-   * Verifies that a response with {@code httpStatusCode} throws {@link
-   * RegistryUnauthorizedException}.
+   * Verifies that a response with {@code httpStatusCode} throws
+   * {@link RegistryUnauthorizedException}.
    */
   private void verifyThrowsRegistryErrorException(int httpStatusCode)
       throws IOException, RegistryException {
@@ -489,21 +482,17 @@ public class RegistryEndpointCallerTest {
       Assert.fail("Call should have failed");
 
     } catch (RegistryErrorException ex) {
-      Assert.assertThat(
-          ex.getMessage(),
-          CoreMatchers.startsWith(
-              "Tried to actionDescription but failed because: unknown error code: code (message)"));
+      Assert.assertThat(ex.getMessage(), CoreMatchers.startsWith(
+          "Tried to actionDescription but failed because: unknown error code: code (message)"));
     }
   }
 
-  
+
   private void setUpRegistryResponse(Exception exceptionToThrow)
       throws MalformedURLException, IOException {
-    Mockito.when(
-            mockHttpClient.call(
-                Mockito.eq("httpMethod"),
-                Mockito.eq(new URL("https://serverUrl/v2/api")),
-                Mockito.any()))
+    Mockito
+        .when(mockHttpClient.call(Mockito.eq("httpMethod"),
+            Mockito.eq(new URL("https://serverUrl/v2/api")), Mockito.any()))
         .thenThrow(exceptionToThrow);
   }
 }

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -87,8 +87,8 @@ public class RegistryEndpointCallerTest {
     @Nullable
     @Override
     public String handleResponse(Response response) throws IOException {
-      return CharStreams
-          .toString(new InputStreamReader(response.getBody(), StandardCharsets.UTF_8));
+      return CharStreams.toString(
+          new InputStreamReader(response.getBody(), StandardCharsets.UTF_8));
     }
 
     @Override
@@ -109,23 +109,24 @@ public class RegistryEndpointCallerTest {
     return mock;
   }
 
-  @Rule
-  public final RestoreSystemProperties systemPropertyRestorer = new RestoreSystemProperties();
+  @Rule public final RestoreSystemProperties systemPropertyRestorer = new RestoreSystemProperties();
 
-  @Mock
-  private EventHandlers mockEventHandlers;
-  @Mock
-  private FailoverHttpClient mockHttpClient;
-  @Mock
-  private Response mockResponse;
+  @Mock private EventHandlers mockEventHandlers;
+  @Mock private FailoverHttpClient mockHttpClient;
+  @Mock private Response mockResponse;
 
   private RegistryEndpointCaller<String> endpointCaller;
 
   @Before
   public void setUp() throws IOException {
-    endpointCaller = new RegistryEndpointCaller<>(mockEventHandlers, "userAgent",
-        new TestRegistryEndpointProvider(), Authorization.fromBasicCredentials("user", "pass"),
-        new RegistryEndpointRequestProperties("serverUrl", "imageName"), mockHttpClient);
+    endpointCaller =
+        new RegistryEndpointCaller<>(
+            mockEventHandlers,
+            "userAgent",
+            new TestRegistryEndpointProvider(),
+            Authorization.fromBasicCredentials("user", "pass"),
+            new RegistryEndpointRequestProperties("serverUrl", "imageName"),
+            mockHttpClient);
 
     Mockito.when(mockResponse.getBody())
         .thenReturn(new ByteArrayInputStream("body".getBytes(StandardCharsets.UTF_8)));
@@ -248,8 +249,9 @@ public class RegistryEndpointCallerTest {
 
     } catch (IOException ex) {
       Assert.assertSame(ioException, ex);
-      Mockito.verify(mockEventHandlers).dispatch(
-          LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
+      Mockito.verify(mockEventHandlers)
+          .dispatch(
+              LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    java.io.IOException\u001B[0m"));
       Mockito.verify(mockEventHandlers)
@@ -269,17 +271,19 @@ public class RegistryEndpointCallerTest {
 
     } catch (IOException ex) {
       Assert.assertSame(ioException, ex);
-      Mockito.verify(mockEventHandlers).dispatch(
-          LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
+      Mockito.verify(mockEventHandlers)
+          .dispatch(
+              LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    java.io.IOException\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    this is due to broken pipe\u001B[0m"));
       Mockito.verify(mockEventHandlers)
-          .dispatch(LogEvent.error(
-              "\u001B[31;1mbroken pipe: the server shut down the connection. Check the server "
-                  + "log if possible. This could also be a proxy issue. For example, a proxy "
-                  + "may prevent sending packets that are too large.\u001B[0m"));
+          .dispatch(
+              LogEvent.error(
+                  "\u001B[31;1mbroken pipe: the server shut down the connection. Check the server "
+                      + "log if possible. This could also be a proxy issue. For example, a proxy "
+                      + "may prevent sending packets that are too large.\u001B[0m"));
       Mockito.verifyNoMoreInteractions(mockEventHandlers);
     }
   }
@@ -293,8 +297,9 @@ public class RegistryEndpointCallerTest {
       Assert.fail();
 
     } catch (IOException ex) {
-      Mockito.verify(mockEventHandlers).dispatch(
-          LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
+      Mockito.verify(mockEventHandlers)
+          .dispatch(
+              LogEvent.error("\u001B[31;1mI/O error for image [serverUrl/imageName]:\u001B[0m"));
       Mockito.verify(mockEventHandlers)
           .dispatch(LogEvent.error("\u001B[31;1m    java.io.IOException\u001B[0m"));
       Mockito.verify(mockEventHandlers)
@@ -398,8 +403,9 @@ public class RegistryEndpointCallerTest {
   @Test
   public void testNewRegistryErrorException_jsonErrorOutput() {
     ResponseException httpException = Mockito.mock(ResponseException.class);
-    Mockito.when(httpException.getContent()).thenReturn(
-        "{\"errors\": [{\"code\": \"MANIFEST_UNKNOWN\", \"message\": \"manifest unknown\"}]}");
+    Mockito.when(httpException.getContent())
+        .thenReturn(
+            "{\"errors\": [{\"code\": \"MANIFEST_UNKNOWN\", \"message\": \"manifest unknown\"}]}");
 
     RegistryErrorException registryException =
         endpointCaller.newRegistryErrorException(httpException);
@@ -423,11 +429,11 @@ public class RegistryEndpointCallerTest {
     Assert.assertEquals(
         "Tried to actionDescription but failed because: registry returned error code 404; "
             + "possible causes include invalid or wrong reference. Actual error output follows:\n"
-            + ">>>>> (404) page not found <<<<<\n" + " | If this is a bug, please file an issue at "
+            + ">>>>> (404) page not found <<<<<\n"
+            + " | If this is a bug, please file an issue at "
             + "https://github.com/GoogleContainerTools/jib/issues/new",
         registryException.getMessage());
   }
-
 
   @Test
   public void testNewRegistryErrorException_nullJsonErrorOutput() {
@@ -448,8 +454,8 @@ public class RegistryEndpointCallerTest {
   }
 
   /**
-   * Verifies that a response with {@code httpStatusCode} throws
-   * {@link RegistryUnauthorizedException}.
+   * Verifies that a response with {@code httpStatusCode} throws {@link
+   * RegistryUnauthorizedException}.
    */
   private void verifyThrowsRegistryUnauthorizedException(int httpStatusCode)
       throws IOException, RegistryException {
@@ -467,8 +473,8 @@ public class RegistryEndpointCallerTest {
   }
 
   /**
-   * Verifies that a response with {@code httpStatusCode} throws
-   * {@link RegistryUnauthorizedException}.
+   * Verifies that a response with {@code httpStatusCode} throws {@link
+   * RegistryUnauthorizedException}.
    */
   private void verifyThrowsRegistryErrorException(int httpStatusCode)
       throws IOException, RegistryException {
@@ -482,17 +488,20 @@ public class RegistryEndpointCallerTest {
       Assert.fail("Call should have failed");
 
     } catch (RegistryErrorException ex) {
-      Assert.assertThat(ex.getMessage(), CoreMatchers.startsWith(
-          "Tried to actionDescription but failed because: unknown error code: code (message)"));
+      Assert.assertThat(
+          ex.getMessage(),
+          CoreMatchers.startsWith(
+              "Tried to actionDescription but failed because: unknown error code: code (message)"));
     }
   }
 
-
   private void setUpRegistryResponse(Exception exceptionToThrow)
       throws MalformedURLException, IOException {
-    Mockito
-        .when(mockHttpClient.call(Mockito.eq("httpMethod"),
-            Mockito.eq(new URL("https://serverUrl/v2/api")), Mockito.any()))
+    Mockito.when(
+            mockHttpClient.call(
+                Mockito.eq("httpMethod"),
+                Mockito.eq(new URL("https://serverUrl/v2/api")),
+                Mockito.any()))
         .thenThrow(exceptionToThrow);
   }
 }

